### PR TITLE
Blocking metrics

### PR DESCRIPTION
**Pull request by zschira**
## Background

This PR adds the initial infrastructure for experiment tracking. It uses MLflow to record metrics from the blocking step, and associate these metrics with the model configuration. At first I was hesitant to use MLflow because I didn't want to have to setup a remote tracking server due to the upfront work required, and added complexity during development. I also wasn't sure that we'd get a ton of value out of MLflow locally, but now I think we can likely make it work. I've decided (at least for the time being) that we can make use of MLflow in a mostly local setup.

There are a few main parts of this PR:

`config.py` - Here I've tried to take any inputs that feed into the blocking step and define them in pydantic models. This allows the model configuration to be easily serialized and saved as json files with the experiment tracking. It will also be really useful after dagsterization, but these config models can feed directly into dagster assets.

`metrics.py` - This module contains a single function that computes and logs metrics for the blocking step. By default it will store the results in a directory called `mlruns` in the base of this repo.

`blocking_experiments.ipynb` - Largely unchanged, but it will now use the function in `metrics.py` to record results from the run.

### Executing a run
Right now a blocking experiment run can be executed/logged using the `measure_blocking` function. This happens in `blocking_experiments.ipynb`, but can also be kicked off directly from a python script. As we add the `splink` layer or test other models we can add more experiments.

### Viewing results
Using the command `mlflow ui` from the base of this repo (or wherever your `mlruns` directory lives) will allow you to visualize and query runs.

### CI integration
I think an important use case is to be able to compute and compare metrics in the CI. My vision to accomplish this without needing a remote tracking server is to add a github action that will run all experiments every time there's a push to `main`, and cache the `mlruns` directory on the runner. Then, anytime there's a PR to main, we can restore the cache and compare the metrics from the latest commit to `main` against those from the feature branch.

**Steps:**

- [x] Add script to `metrics.py` to execute a set of desired experiments
- [x] Create action on `push` to run `metrics.py` script and cache results
- [x] Add step to `tox-pytest` to compare metrics and attach results to PR in the form of a nice table

### CODE REVIEW

File: src/ferc1_eia_match/config.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    start_year: int
+    end_year: int
+
+
+class ColumnEmbedding(BaseModel):
+    """Specify embedding vectorizer to use for column and any options."""
+
+    embedding_type: str
+    options: dict = {}
```
Comment (katie-lamb) [inline src/ferc1_eia_match/config.py:17]: I'm interpreting `options` as potential keyword arguments to the embedding function specified by `embedding_type`? I'm going to just use the `config.embedding.embedding_map` as the input to the DataFrameEmbedder.

**Comment by katie-lamb** [comment]
This looks good so far! I updated the `blocking_experiments` notebook to use `model_config` as inputs to various functions instead of redefining each dictionary.

File: src/ferc1_eia_match/config.py (hunk 1 above)
Comment (zschira) [reply src/ferc1_eia_match/config.py:17]: Yeah that was the plan. And that makes sense, should be cleaner that way

**Comment by zschira** [comment]
@katie-lamb this should be ready for a full review now.

I've added two new workflows, `compare-metrics` and `compute-metrics`. `compare-metrics` will run on a PR and should attach a markdown report to the PR with experiment results from the base (latest commit on `main`) and results from the feature branch. `compute-metrics` will run anytime new commits get pushed to main and compute experiment results for that commit, then cash the results to be used in `compute-metrics`.

**Comment by zschira** [comment]
Did some debugging on the `compare-metrics` action, and solved a number of little problems with it. I'm pretty sure it's only failing now due to memory issues when generating the plant parts list, so it's probably best to wait for that to be fully dagsterized before spending more time trying to get it running.

**Review by katie-lamb — COMMENTED** [review]
This is great, I'm excited to get this in action. I'm no longer seeing the results of my experiment show up in `mlruns` when I run `metrics.blocking.measure_blocking` from the `blocking_experiments` notebook. Is that intended so that the only metrics in `mlruns` are those that are computed when there are commits to `main` (the new baseline experiment)? I'm seeing an old run that I did in there but maybe with the new changes not all runs will be cached in `mlruns`?

From my understanding the workflow is like this: 
- make changes/make experiment config and run `measure_blocking` to see some metrics
- commit changes and make a PR, `compare-metrics` will run so you can see the new changes in comparison to the cached baseline on `main`
- when commits are made to `main` the metrics will get saved in `mlruns` and the latest commit will be the baseline

**Comment by zschira** [comment]
> This is great, I'm excited to get this in action. I'm no longer seeing the results of my experiment show up in mlruns when I run metrics.blocking.measure_blocking from the blocking_experiments notebook. Is that intended so that the only metrics in mlruns are those that are computed when there are commits to main (the new baseline experiment)? I'm seeing an old run that I did in there but maybe with the new changes not all runs will be cached in mlruns?

I think you probably have 2 `mlruns` directories, one in `notebooks`, and one in the base of your repo. I've updated the notebook, so it will tell `measure_blocking` to use the dir at the base of the repo by default. Let me know if that's not the case though.

> make changes/make experiment config and run measure_blocking to see some metrics

Yeah sorry this is a bit confusing, but `measure_blocking` is what you can use during development and can be called from anywhere to test different configs/refine the model. `execute_blocking` (now `run_blocking_tests`) is used by the CI via the `compute_metrics` script as a sort of repeatable test that handles all of the setup, but is using `measure_blocking` underneath.

> commit changes and make a PR, compare-metrics will run so you can see the new changes in comparison to the cached baseline on main

  >  I shouldn't be seeing experiment_results.md for this PR right?

> when commits are made to main the metrics will get saved in mlruns and the latest commit will be the baseline
Yeah this is all correct, when new commits get to `main`, they'll become the baseline that will be used in CI comparisons, but for now this is failing due to memory issues when generating the plant parts list, so `experiment_results.md` isn't getting generated, but it would otherwise.

File: src/ferc1_eia_match/metrics/blocking.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        menu_blocks_dict=embedder.right_blocks_dict,
+    )
+
+    # read in training data
+    pkg_source = importlib.resources.files("ferc1_eia_match.package_data").joinpath(
+        "ferc1_eia_train.csv"
+    )
+    with importlib.resources.as_file(pkg_source) as csv_file:
+        train_df = pd.read_csv(csv_file)
+    ks = [5, 10, 15, 20, 25, 30, 40, 50]
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/ferc1_eia_match/metrics/blocking.py]: Yeah that seems reasonable, I've updated this to accept `ks` as a kwarg

File: src/ferc1_eia_match/metrics/compute.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+def main():
+    """Execute experiments."""
+    args = _parse_args()
+    pudl_engine = sa.create_engine(args.db)
+
+    for experiment in args.experiments:
+        match experiment:
+            case "blocking":
+                blocking.execute_blocking(pudl_engine, Path(args.mlrun_uri))
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/ferc1_eia_match/metrics/compute.py]: Yeah `compute.py` and `compare.py` are really there for the CI. I have updated it though so you can pass a JSON file or dict for configuration. When running from the notebook, though, all you need is `measure_blocking`.

File: src/ferc1_eia_match/metrics/blocking.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            # Compute % that predict match as first value in candidate set
+            first_match_is_correct = np.in1d(
+                eia_train_idx, candidate_set[ferc_train_idx][:, 0]
+            )
+            n_first_match_correct = np.sum(first_match_is_correct)
+            mlflow.log_metric(
+                "percent_first_match", n_first_match_correct / len(train_df_with_idx)
+            )
+            logger.info(
+                f"k: {k}, metric: {metric} percent_first: {n_first_match_correct/len(train_df_with_idx)}"
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/ferc1_eia_match/metrics/blocking.py:88]: Yeah I guess it's not super useful, but maybe there's value in being able to see it without running a test with `k=1`. I don't have particularly strong feelings either way.

**Review by katie-lamb — APPROVED** [review]
This all works well for me and looks good!

**Pull request merged**